src/main/java/org/truffleruby/core/array/ArrayEachIteratorNode.java

```diff
@@ -42,36 +42,17 @@ public static ArrayEachIteratorNode create() {
     public abstract RubyArray execute(RubyArray array, RubyProc block, int startAt,
             ArrayElementConsumerNode consumerNode);
 
-    @Specialization(
-            guards = { "array.size == 1", "startAt == 0" },
-            limit = "storageStrategyLimit()")
-    protected RubyArray iterateOne(RubyArray array, RubyProc block, int startAt, ArrayElementConsumerNode consumerNode,
-            @Bind("array.store") Object store,
-            @CachedLibrary("store") ArrayStoreLibrary stores) {
-
-        consumerNode.accept(array, block, stores.read(store, 0), 0);
-
-        if (array.size > 1) {
-            // Implicitly profiles through lazy node creation
-            return getRecurseNode().execute(array, block, 1, consumerNode);
-        }
-
-        return array;
-    }
-
-    @Specialization(
-            guards = { "array.size != 1" },
-            limit = "storageStrategyLimit()")
+    @Specialization(limit = "storageStrategyLimit()")
     protected RubyArray iterateMany(RubyArray array, RubyProc block, int startAt, ArrayElementConsumerNode consumerNode,
-            @Bind("array.store") Object store,
```

Comment:
I think you need to keep this `@Bind`.

Reply:
This was actually the bug here, in general `@Bind` is useful to avoid the store changing, and `@CachedLibrary("array.store")` no longer accepting it.
But here we check `stores.accepts(array.store))` below, so it's not a problem.

Reply:
I reviewed other usages of `accepts()` + `@Bind` and they seem fine.